### ISSUE TITLE
3.7 Landmark Classification & Tagging for Social Media

### ISSUE THREAD

**Issue by Alex1986ham** [Alex1986ham/data_lakehouse_with_spark#8]
In this project, you will apply the skills you have acquired in the Convolutional Neural Network (CNN) course to build a landmark classifier.

Photo sharing and photo storage services like to have location data for each photo that is uploaded. With the location data, these services can build advanced features, such as automatic suggestion of relevant tags or automatic photo organization, which help provide a compelling user experience. Although a photo's location can often be obtained by looking at the photo's metadata, many photos uploaded to these services will not have location metadata available. This can happen when, for example, the camera capturing the picture does not have GPS or if a photo's metadata is scrubbed due to privacy concerns.

If no location metadata for an image is available, one way to infer the location is to detect and classify a discernible landmark in the image. Given the large number of landmarks across the world and the immense volume of images that are uploaded to photo sharing services, using human judgment to classify these landmarks would not be feasible.

In this project, you will take the first steps towards addressing this problem by building models to automatically predict the location of the image based on any landmarks depicted in the image. You will go through the machine learning design process end-to-end: performing data preprocessing, designing and training CNNs, comparing the accuracy of different CNNs, and deploying an app based on the best CNN you trained.

![](https://video.udacity-data.com/topher/2021/February/602dac82_landmarks-example/landmarks-example.png)

Examples from the landmarks dataset: a road in Death Valley, the Brooklyn Bridge, and the Eiffel Tower

The high level steps of the project include:

Create a CNN to Classify Landmarks (from Scratch) - Here, you'll visualize the dataset, process it for training, and then build a convolutional neural network from scratch to classify the landmarks. You'll also describe some of your decisions around data processing and how you chose your network architecture. You will then export your best network using Torch Script.
Create a CNN to Classify Landmarks (using Transfer Learning) - Next, you'll investigate different pre-trained models and decide on one to use for this classification task. Along with training and testing this transfer-learned network, you'll explain how you arrived at the pre-trained network you chose. You will also export your best transfer learning solution using Torch Script
Deploy your algorithm in an app - Finally, you will use your best model to create a simple app for others to be able to use your model to find the most likely landmarks depicted in an image. You'll also test out your model yourself and reflect on the strengths and weaknesses of your model.
Each of these three major steps is carried out in a corresponding Jupyter Notebook that you will be provided in the project starter kit. So there are three notebooks that will guide you in your project work.
Starter Code and Instructions
Starter code is provided to you in the project workspace, and is also available for you to download if you prefer to work locally. Please follow the detailed instructions contained in the following three notebooks:

cnn_from_scratch.ipynb: Create a CNN from scratch
transfer_learning.ipynb: Use transfer learning
app.ipynb: Deploy your best model in an app. At the end of this notebook you will also generate the archive file that you will submit for review
Evaluation
Your project will be reviewed by a Udacity reviewer against the [project rubric(opens in a new tab)](https://learn.udacity.com/rubric/4814). Review this rubric thoroughly, and evaluate your project against the rubric before submission. Your project must satisfy every criterion in the rubric for you to pass.

We have also written some tests for each step of the process to help you self-diagnose the most common problems. You will be running these tests as part of the development process, as described in the notebooks. Do not skip ahead if the tests are not passing. When tests fail, they will usually provide some clear guidance on what went wrong.

Project Submission
At the end of the app.ipynb notebook there are instructions on how to generate a file with a name like submission_2022-03-31T00h55m.tar.gz that contains all your notebooks and code. If you worked in the Project Workspace, you will download that file from the workspace to your machine. If you worked locally the file will already be on your machine, in the same directory as the notebooks. After you have obtained the file, go to the Project Submission page and submit it.

For this project, you can either complete the work on your own local computer or you can complete it within a workspace right here in the Udacity classroom.

Option 1: Working in the Udacity project workspace
You can find the Udacity Project Workspace here in your Udacity classroom, on the Project Workspace page.

On the workspace page, when prompted on whether you want a GPU or not, please ANSWER YES (the GPU is going to make everything several times faster).

The environment is already set up for you, including the starter code, so you can jump right into building your project!
Option 2: Working locally
Alternatively, you can develop your project locally on your computer. This setup requires a bit of familiarity with creating a working deep learning environment. While things should work out of the box, in case of problems you might have to do operations on your system (like installing new NVIDIA drivers for your GPU) that are not covered in the class. Please work on your local environment only if you are at least a bit familiar with these subjects, otherwise we recommend using the provided Udacity workspace in the classroom.

Download the starter kit from the Resources section of the left sidebar of the Udacity classroom here. (Scroll down if you don't see it.)
Open a terminal and navigate to the directory where you installed the starter kit.
Download and install [Miniconda(opens in a new tab)](https://docs.conda.io/en/latest/miniconda.html)
Create a new conda environment with Python 3.7.6:
conda create --name udacity python=3.7.6 pytorch=1.11.0 torchvision torchaudio cudatoolkit -c pytorch
Activate the environment:
conda activate udacity
NOTE: You will need to activate your environment again every time you open a new terminal.

Install the required packages for the project:
pip install -r requirements.txt
Test that the GPU is working (execute this only if you have a NVIDIA GPU on your machine, which Nvidia drivers properly installed)
python -c "import torch;print(torch.cuda.is_available())
This should return True. If it returns False your GPU cannot be recognized by pytorch. Test with nvidia-smi that your GPU is working. If it is not, check your NVIDIA drivers. If you cannot solve this problem, please consider using the Project Workspace instead of working locally.

Install and open jupyter lab:
pip install jupyterlab 
jupyter lab

Environment and Dependencies (for local development)
If you prefer to work on your own machine rather than use the Project Workspace here in the classroom, make sure you have everything listed below.

If you are experiencing issues with the workspace, try these troubleshooting tips:

Refresh the workspace using the Menu option in the bottom left of the workspace
After downloading any scripts or work you've done, Reset data in the workspace using the Menu option in the bottom left of the workspace.
Restart the kernel after the !pip install cell in the notebook using the Kernel menu at the top of the Jupyter Notebook in the menu bar.
Run the!pip install cell, and setup_env()cell every time you return to the workspace and it starts. Make sure the setup_env() completes before you attempt to call Pytest in the cells after setup_env().

Suggestions to Make Your Project Stand Out

Keep iterating on your model and training parameters to see how high you can get your test accuracy! For the architecture from scratch, you should be able to get above 60%. For the transfer learning, with the proper architecture you can reach 80% or more. One example idea to experiment with is to apply different augmentations to your training data. You can also try different architectures, try to avoid overfitting by increasing dropout and/or regularization (weight_decay), experiment with different batch sizes…. Keep track of your experiments and provide a table with your train loss, validation loss, accuracy, and notes about what you think is going on (for example, “the model is overfitting on the training set because the validation loss stops decreasing while the train loss keeps decreasing). Note the experiment that gave you the best accuracy. NOTE: do NOT look at the performance on the test set when you optimize your hyperparameters, only optimize on the validation set. Then when you have converged on the best parameters, run on the test set and verify that the performance there is similar to what you got on the validation set.
Use the features from the penultimate layer of your from-scratch CNN or transfer-learned CNN to implement an image retrieval algorithm. Your algorithm should roughly perform the following procedure: given an image, extract the CNN features for the image, compute the dot product between the aforementioned CNN features and the CNN features for each of the images in landmark_images, return the images that have the highest dot product values.
Include in your submission a discussion around additional use cases of your model - what other situations might it be useful?

Workbook: cnn_from_scratch.ipynb
